cpp/src/parquet/encryption/encryption_internal.cc

```diff
@@ -463,10 +463,13 @@ int AesDecryptor::AesDecryptorImpl::GcmDecrypt(const uint8_t* ciphertext,
       throw ParquetException("Wrong ciphertext length");
     }
     ciphertext_len = written_ciphertext_len + length_buffer_length_;
-  } else {
-    if (ciphertext_len == 0) {
-      throw ParquetException("Zero ciphertext length");
-    }
+  }
+
+  if (ciphertext_len < length_buffer_length_ + kNonceLength + kGcmTagLength) {
+    std::stringstream ss;
+    ss << "Invalid ciphertext length " << ciphertext_len << ". Expected at least "
+       << length_buffer_length_ + kNonceLength + kGcmTagLength << "\n";
+    throw ParquetException(ss.str());
   }
```

Comment:
Looking at the existing code in these methods, several things stand out:
1) we don't validate ciphertext length before fetching the 4 bytes encoding the actual length (a corrupt file could perhaps have a ciphertext length < 4?)
2) use of raw C arrays instead of `std::array<uint8_t>` for example
3) why is the `ciphertext_len` argument optional in this API? this looks fickle and error-prone.
4) the part that extracts and validates the actual length is duplicated in the two methods

I would suggest we take the opportunity and refactor this into a cleaner and less error-prone implementation. In particular, the GCM and CTR-specific methods should probably have a mandatory ciphertext length, and would not have to bother with reading the length bytes.

@ggershinsky @thamht4190 Do we have an explanation for the very odd choices here?

Reply:
Yeah I didn't dig in too deep to find why the `ciphertext_len` is optional, it would be nice if that could be mandatory. But if that's not possible we should at least be able to provide the size of the buffer that the ciphertext is being read from to ensure that `ciphertext_len` isn't greater than this.

Reply:
It does look like we don't always know the length of the ciphertext but sometimes just an upper bound that's used to allocate the buffer. I found an example of this when reading the bloom filter header:
https://github.com/apache/arrow/blob/e615a3034ed70c9bdbc0b74882f7345c0c2b8912/cpp/src/parquet/bloom_filter.cc#L116-L121

I'm thinking that rather than having separate arguments like `ciphertext_buffer_len` (required) and `ciphertext_expected_len` (optional), it's probably fine to make `ciphertext_len` required and mean the size of the buffer, so we would validate that the actual length is <= this after accounting for the 4 byte length rather than enforcing an exact match. Does that seem reasonable?